tpm2/marshalling.go

```diff
@@ -126,3 +126,85 @@ func (b *boxed[T]) unmarshal(buf *bytes.Buffer) error {
 	b.Contents = new(T)
 	return unmarshal(buf, reflect.ValueOf(b.Contents))
 }
+
+// MarshalCommandResponse marshals both command and response.
+func MarshalCommandResponse[C Command[R, *R], R any](cmd C, rsp *R) (cmdData []byte, rspData []byte, err error) {
+	cmdData, err = MarshalCommand(cmd)
+	if err != nil {
+		return nil, nil, fmt.Errorf("marshalling command: %w", err)
+	}
+	rspData, err = MarshalResponse(rsp)
+	if err != nil {
+		return nil, nil, fmt.Errorf("marshalling response: %w", err)
+	}
+	return cmdData, rspData, nil
+}
+
+// UnmarshalCommandResponse unmarshals both command and response.
+func UnmarshalCommandResponse[C Command[R, *R], R any](cmdData []byte, rspData []byte) (cmd C, rsp *R, err error) {
+	cmd, err = UnmarshalCommand[C, R](cmdData)
+	if err != nil {
+		return cmd, rsp, fmt.Errorf("unmarshalling command: %w", err)
+	}
+	rsp, err = UnmarshalResponse[R](rspData)
+	if err != nil {
+		return cmd, rsp, fmt.Errorf("unmarshalling response: %w", err)
+	}
+	return cmd, rsp, nil
+}
+
+// MarshalCommand marshals a TPM command.
+func MarshalCommand[C Command[R, *R], R any](cmd C) ([]byte, error) {
+	var buf bytes.Buffer
+	params := taggedMembers(reflect.ValueOf(cmd), "handle", true)
+	for i := range len(params) {
+		if err := marshalParameter(&buf, cmd, i); err != nil {
+			return nil, fmt.Errorf("marshalling command's parameter: %w", err)
+		}
+	}
+	return buf.Bytes(), nil
```

Comment:
In #414 we discussed how the audit verifier, who needs to verify `cpHash` (the command parameter hash), needs not only the command parameters, but also the names. Instead of returning just `parameters`, should we consider returning a buffer that directly hashes to `cpHash`? That buffer is ( commandCode || name1 [ || name2 [ || name3 ]] || parms ).

See CPHash, which I think is similar to what you need (except obviously that it returns the digest when you want the preimage): https://github.com/google/go-tpm/blob/bf120203f23af1e4291a336cb93a7a79e2a0832f/tpm2/sessions.go#L56

Reply:
If we chose the marshaled ( commandCode || name1 [ || name2 [ || name3 ]] || parms ) as the thing to be returned by this function, it would be neat to provide an Unmarshal function that takes the whole thing and returns (TPM_CC, TPM2B_Name[], Command, error). Since the CC is there, we can easily look up how many names to expect and fail fast & obvious when there is a mismatch.